[qpack] Overhaul the pseudocode (#3577)
quicwg/base-drafts: draft-ietf-quic-qpack.md

# Sample One Pass Encoding Algorithm

Pseudo-code for single pass encoding, excluding handling of duplicates,
non-blocking mode, available encoder stream flow control and reference tracking.

~~~
base = dynamicTable.getInsertCount()
requiredInsertCount = 0
for header in headers:

>   staticIndex = staticTable.getIndex(header)

Reviewer: ```suggestion
  staticIndex = staticTable.findIndex(header)
```

This is a search, right?

  if staticIndex is not None:
    encodeIndexReference(streamBuffer, staticIndex)
    continue

>   dynamicIndex = dynamicTable.getIndex(header)

Reviewer: ```suggestion
  dynamicIndex = dynamicTable.findIndex(header)
```

  if dynamicIndex is None:
    # No matching entry.  Either insert+index or encode literal

>     # getNameIndex attempts to find an index for this header's name.
>     # If one exists, it returns the name's (absolute) index and a

Reviewer 2: ```suggestion
    # getNameIndex searches for this header's name in the table.
    # If an entry matches, it returns the (absolute) index and a
```

    # boolean indicating which table has the name.  If the name
    # can't be found, nameIndex is None

>     nameIndex, isStaticName = getNameIndex(header.name)

Reviewer: This is a little strange, as you are relying on functions on the tables to find the full entry, but you jump to a different approach for looking up the name.  I realize that two lookups are worse than one, but it's an odd stylistic jump, that's all.

```python
# An alternative might be to do the name lookup like this:
staticNameIndex = staticTable.findName(header.name)
if staticNameIndex is None:
    dynamicNameIndex = dynamicTable.findName(header.name)
if header.canIndex() and dynamicTable.canAdd(header):
    encodeInsert(encoderBuffer, staticNameIndex, dynamicNameIndex, header)
    dynamicIndex = dynamicTable.add(header)

    # For a bonus, continuing from the above, forget about the
    # newly inserted dynamic table entry if we are at the limit of blocked streams.
    if numBlockedStreams >= maxBlockedStreams:
        dynamicIndex = None
    else:
        addedBlockingEntry = True
        # Later, you can increase numBlockedStreams, but you only do that once.
```

    if shouldIndex(header) and dynamicTable.canIndex(header):

>       encodeInsert(encoderBuffer, nameIndex, isStaticName, header)

Reviewer 2: ```suggestion
    encodeInsert(
        encoderBuffer, nameIndex, isStaticName, header
    )
```

      dynamicTable.add(header)

>       dynamicIndex = dynamicTable.getInsertCount()

Reviewer: I can't suggest here, but I think that you just want this:

```python
        dynamicIndex = dynamicTable.add(header)
```

Otherwise you are breaking encapsulation in a small way: relying on the insert count for getting the index of the thing you just added.

  if dynamicIndex is None:
    # Couldn't index it, literal
    if nameIndex is None or isStaticName:
      # Encodes a literal with a static name or literal name
      encodeLiteral(streamBuffer, nameIndex, header)
    else:
      # encode literal with dynamic name, possibly above base
      encodeDynamicLiteral(streamBuffer, nameIndex, base, header)
      requiredInsertCount = max(requiredInsertCount, nameIndex)
  else:
    # Dynamic index reference
    assert(dynamicIndex is not None)
    requiredInsertCount = max(requiredInsertCount, dynamicIndex)
    # Encode dynamicIndex, possibly above base
    encodeDynamicIndexReference(streamBuffer, dynamicIndex, base)

# encode the prefix
if requiredInsertCount == 0:
  encodeIndexReference(prefixBuffer, 0, 0, 8)
  encodeIndexReference(prefixBuffer, 0, 0, 7)
else:

>   wireRIC =
>     (requiredInsertCount % (2 * getMaxEntries(maxTableCapacity))) + 1;

Reviewer 2: ```suggestion
  wireRIC = (
    requiredInsertCount
    % (2 * getMaxEntries(maxTableCapacity))
  ) + 1
```
